node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1441,29 +1419,73 @@ async fn fetch_finality_signatures_by_block_hash(
-                    ?block_hash,
+                    ?block_header_hash,
                     ?peer,
                     "did not get FinalitySignatures from peer, got from storage instead",
                 );
                 Some(*item)
             }
             Ok(FetchedData::FromPeer { item, .. }) => {
-                trace!(?block_hash, ?peer, "fetched FinalitySignatures");
+                trace!(?block_header_hash, ?peer, "fetched FinalitySignatures");
                 Some(*item)
             }
             Err(_) => None,
         };
 
         sig_collector.add_signatures(maybe_signatures);
 
-        if sig_collector
-            .check_if_sufficient(validator_weights, ctx.config.finality_threshold_fraction())
-        {
-            break;
+        let era_for_validators_retrieval =
+            get_era_id_for_validators_retrieval(&block_header.era_id());
+
+        let maybe_switch_block_of_previous_era = ctx
+            .effect_builder
+            .get_switch_block_header_at_era_id_from_storage(era_for_validators_retrieval)
+            .await;
+
+        match maybe_switch_block_of_previous_era {
+            Some(switch_block_of_previous_era) => {
+                match switch_block_of_previous_era.next_era_validator_weights() {
+                    Some(validator_weights) => {
+                        if sig_collector.check_if_sufficient(
+                            validator_weights,
+                            ctx.config.finality_threshold_fraction(),
+                        ) {
+                            info!(
+                                ?block_header_hash,
+                                height = block_header.height(),
+                                ?era_for_validators_retrieval,
+                                "fetched sufficient finalty signatures"
+                            );
+                            break;
+                        }
+                    }
+                    None => {
+                        return Err(Error::HitGenesisBlockTryingToGetTrustedEraValidators {
+                            trusted_header: block_header,
+                        });
+                    }
+                }
+            }
+            None => {
+                return Err(Error::HitGenesisBlockTryingToGetTrustedEraValidators {
+                    trusted_header: block_header,
+                });
+            }
         }
     }
+    Ok(sig_collector.into_inner())
+}
 
-    sig_collector.into_inner()
+/// Returns the EraId whose switch block should be used to obtain validator weights.
+fn get_era_id_for_validators_retrieval(era_id: &EraId) -> EraId {
+    if *era_id != EraId::from(0) {
+        // For eras > 0 we need to use validator set from the previous era.
+        *era_id - 1
+    } else {
+        // In case of Era=0 there's no previous era, but since validators never change during
+        // that era we can safely use the Era 0's switch block.
+        EraId::from(0)
+    }
```

Comment:
When we've decided how to handle that we _might_ have to also return `*era_id` in the emergency restart case here.

Reply:
Fixed in https://github.com/casper-network/casper-node/pull/3125/commits/89c0946603c30390a55b78032ea79c18b67013c3